jetty-io/src/main/java/org/eclipse/jetty/io/DefaultRetainableByteBufferPool.java

```diff
@@ -0,0 +1,314 @@
+    public long getDirectMemory()
+    {
+        return getMemory(true);
+    }
+
+    @ManagedAttribute("The bytes retained by heap ByteBuffers")
+    public long getHeapMemory()
+    {
+        return getMemory(false);
+    }
+
+    private long getMemory(boolean direct)
+    {
+        if (direct)
+            return _currentDirectMemory.get();
+        else
+            return _currentHeapMemory.get();
+    }
+
+    @ManagedAttribute("The available bytes retained by direct ByteBuffers")
+    public long getAvailableDirectMemory()
+    {
+        return getAvailableMemory(true);
+    }
+
+    @ManagedAttribute("The available bytes retained by heap ByteBuffers")
+    public long getAvailableHeapMemory()
+    {
+        return getAvailableMemory(false);
+    }
+
+    private long getAvailableMemory(boolean direct)
+    {
+        Pool<RetainableByteBuffer>[] buckets = direct ? _direct : _indirect;
+        long total = 0L;
+        for (int i = 0; i < buckets.length; i++)
+        {
+            Pool<RetainableByteBuffer> bucket = buckets[i];
+            long capacity = (i + 1L) * _factor;
+            total += bucket.values().stream().filter(Pool.Entry::isIdle).count() * capacity;
+        }
+        return total;
+    }
+
+    @ManagedOperation(value = "Clears this RetainableByteBufferPool", impact = "ACTION")
+    public void clear()
+    {
+        clearArray(_direct, _currentDirectMemory);
+        clearArray(_indirect, _currentHeapMemory);
+    }
+
+    private void clearArray(Pool<RetainableByteBuffer>[] poolArray, AtomicLong memoryCounter)
+    {
+        for (Pool<RetainableByteBuffer> retainableByteBufferPool : poolArray)
+        {
+            for (Pool<RetainableByteBuffer>.Entry entry : retainableByteBufferPool.values())
+            {
+                entry.remove();
+                memoryCounter.addAndGet(-entry.getPooled().capacity());
+            }
+        }
+    }
+
+    private void releaseExcessMemory(boolean direct)
+    {
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (maxMemory > 0)
+        {
+            long excess = getMemory(direct) - maxMemory;
+            if (excess > 0)
+                evict(direct, excess);
+        }
+    }
+
+    /**
+     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
+     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
+     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
+     */
+    private void evict(boolean direct, long excess)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
+        long now = System.nanoTime();
+        long totalClearedCapacity = 0L;
+
+        Pool<RetainableByteBuffer>[] buckets = direct ? _direct : _indirect;
+
+        while (totalClearedCapacity < excess)
```

Comment:
If multiple threads acquire on a full pool, will multiple threads end up iterating here?   Is there a way to ensure only a single thread does evict? or is it not worth worrying about?

Reply:
We discussed this with @sbordet and out conclusion was that it's preferable to have concurrent threads iterating rather than leaving that job to a single one.

The reasoning was that if we end up in a situation where a few threads have to evict concurrently every other day, the impact is going to be negligible no matter how inefficient the eviction may be. But if we are in a situation where eviction happens frequently, this is probably a sign that the pool is undersized and the best thing to do is probably to generate some form of back pressure by making all threads "waste" cycles in the eviction code.